react/features/base/avatar/components/AbstractAvatar.js

```diff
@@ -0,0 +1,215 @@
+// @flow
+
+import { getGravatarURL } from 'js-utils/avatar';
+import { PureComponent } from 'react';
+
+import { getAvatarColor, getInitials } from '../functions';
+
+import { getParticipantById } from '../../participants';
+
+export type Props = {
+
+    /**
+     * URI of the avatar, if identifiable.
+     */
+    _avatarURL: ?string,
+
+    /**
+     * URI of the gravatar if identifyable.
+     */
+    _gravatarURL: ?string,
+
+    /**
+     * The string we base the initials on (this is generated from a list of precendences).
+     */
+    _initialsBase: ?string,
+
+    /**
+     * A string to override the initials to generate a color of. This is handy if you don't want to make
+     * the background color match the string that the initials are generated from.
+     */
+    colorBase?: string,
+
+    /**
+     * Display name of the entity to render an avatar for (if any). This is handy when we need
+     * an avatar for a non-participasnt entity (e.g. a recent list item).
+     */
+    displayName?: string,
+
+    /**
+     * The ID of the participant to render an avatar for (if it's a participant avatar).
+     */
+    participantId?: string,
+
+    /**
+     * The size of the avatar.
+     */
+    size: number,
+
+    /**
+     * URI of the avatar, if any.
+     */
+    uri: ?string,
+}
+
+type State = {
+    avatarFailed: boolean,
+    gravatarFailed: boolean
+}
+
+export const DEFAULT_SIZE = 65;
+
+/**
+ * Implements an abstract class to render avatars in the app.
+ */
+export default class AbstractAvatar<P: Props> extends PureComponent<P, State> {
+    /**
+     * Instantiates a new {@code Component}.
+     *
+     * @inheritdoc
+     */
+    constructor(props: P) {
+        super(props);
+
+        this.state = {
+            avatarFailed: false,
+            gravatarFailed: false
+        };
+
+        this._onAvatarLoadError = this._onAvatarLoadError.bind(this);
+        this._onGravatarLoadError = this._onGravatarLoadError.bind(this);
+    }
+
+    /**
+     * Implements {@code Component#componentDidUpdate}.
+     *
+     * @inheritdoc
+     */
+    componentDidUpdate(prevProps: P) {
+        if (prevProps._avatarURL !== this.props._avatarURL
+            || prevProps._gravatarURL !== this.props._gravatarURL) {
+
+            // URLS changed, so we need to try to fetch them again.
+            // Eslint doesn't like this statement, but based on the React doc, it's safe if it's
+            // wrapped in a condition: https://reactjs.org/docs/react-component.html#componentdidupdate
+
+            // eslint-disable-next-line react/no-did-update-set-state
+            this.setState({
+                avatarFailed: false,
+                gravatarFailed: false
+            });
```

Comment:
Here's an alternative that came to mind and I haven't thought through but want to share: what if there were an image component that took in an array of urls and a placeholder component, and that component knew how to fallback and that component only updated when its props changed. That would eliminate this two-boolean handling.

Reply:
That sounds like a good idea!

Reply:
I like the fact that supporting Gravatar (or any other avatar service) would just be a matter of adding one more URL to the list, making the whole thing more flexible.

Reply:
This went to a new direction, so kinda obsolete now. We now have the array of urls as described above, but there is a prefetch function that does all these and stores the result in Redux, so then we can even signal the avatar change through the external API (this couldn't have been done with a UI element)